### PR TITLE
one by one uses permpal for counting etc

### PR DESCRIPTION
**Pull request by christianbean**

### CODE REVIEW

**Comment by jaypantone** [comment]
Here is a snippet that works on develop but fails on this branch. I'm digging into it further.
```py
import requests
from comb_spec_searcher import CombinatorialSpecification
from comb_spec_searcher.rule_db import RuleDBForest
from tilings.tilescope import TileScope, TileScopePack

uri = "https://permpal.com/perms/raw_data_json/basis/0123_0231"

data = requests.get(uri).json()
spec = CombinatorialSpecification.from_dict(data["specs_and_eqs"][0]["spec_json"])
pack = TileScopePack.from_dict(data["specs_and_eqs"][0]["pack_json"])

T = TileScope(spec.root, pack, ruledb=RuleDBForest())
for tiling in spec.comb_classes():
    T.classqueue.add(T.classdb.get_label(tiling))
spec = T.auto_search()
print(spec.get_maple_equations())
```

**Comment by jaypantone** [comment]
Note that in the snippet above I haven't expanded verified first. What seems to be happening is there is a locally factorable tiling containing an Av(123) cell and it's getting 0,0,0,... as its initial conditions:
```py
[I 230122 11:01:30 specification:385] Computing initial conditions for
+-+-+
| |\|
+-+-+
|1| |
+-+-+
1: Av+(012)
\: Av(01)
Requirement 0:
0: (0, 0)
Assumption 0:
can count points in cell (1, 1)
[I 230122 11:01:30 specification:359] Computing initial conditions
[I 230122 11:01:31 comb_spec_searcher:514] Auto search started
    Initialising CombSpecSearcher for the combinatorial class:
    +-+
    | |
    +-+

    Looking for recursive combinatorial specification with the strategies:
    Inferral: row and column separation, obstruction transitivity
    Initial: factor, point corroboration
    Verification: verify atoms
    Set 1: point insertion, requirement placement, row and column placement

[I 230122 11:01:31 comb_spec_searcher:605] No more classes to expand.
[I 230122 11:01:31 comb_spec_searcher:553] Specification detected.
[I 230122 11:01:31 base:306] Minimizing for 0 seconds.
[I 230122 11:01:31 base:267] Found specification with 1 rules.
[I 230122 11:01:31 comb_spec_searcher:469] Specification built
    Time taken: 0:00:00
    CSS status:
    	Total time accounted for: 0:00:00
                                Number of                                Number of
                             applications    Time spent    Percentage        rules
        -----------------  --------------  ------------  ------------  -----------
        verify atoms                    1       0:00:00           43%            1
        add rule                        1       0:00:00           18%            -
        has specification               1       0:00:00           37%            -

    ClassDB status:
    	Total number of combinatorial classes found is 1
    	is_empty check applied 1 time. Time spent: 0:00:00
    Queue status (currently on level 1):
        Queue              Size
        ---------------  ------
        working               0
        current (set 1)       0
        next                  0
    	The size of the current queues at each level: 1
    RuleDB status:
                                                   Total number
        ---------------------------------------  --------------
        Combinatorial rules                                   1
        Equivalence rules                                     0
        Combintorial rules up to equivalence                  1
        Strategy verified combinatorial classes               1
        Verified combinatorial classes                        1
        combinatorial classes up to equivalence               1
    	Called find equiv path 1 times, for total time of 0.0 seconds.

    Memory Status:
        ---------------------------------  ---------
        OS Allocated                       345.2 MiB
        Current Memory Used                160.8 MiB
        Current Memory Allocated           225.8 MiB
        Current JIT Memory Used              8.5 MiB
        Current JIT Memory Allocated         9.0 MiB
        Peak Memory Used                   263.6 MiB
        Peak Memory Allocated Memory Used  267.8 MiB
        ---------------------------------  ---------
    	Total Garbage Collection Time: 0:00:11
    Specification found has 1 rules
[I 230122 11:01:31 comb_spec_searcher:514] Auto search started
    Initialising CombSpecSearcher for the combinatorial class:
    +-+
    |\|
    +-+
    \: Av(01)
    Assumption 0:
    can count points in cell (0, 0)
    Looking for recursive combinatorial specification with the strategies:
    Inferral: row and column separation, obstruction transitivity
    Initial: factor, point corroboration
    Verification: verify atoms
    Set 1: point insertion, requirement placement, row and column placement

[I 230122 11:01:31 comb_spec_searcher:553] Specification detected.
[I 230122 11:01:31 base:306] Minimizing for 0 seconds.
[I 230122 11:01:31 base:267] Found specification with 4 rules.
[I 230122 11:01:31 comb_spec_searcher:469] Specification built
    Time taken: 0:00:00
    CSS status:
    	Total time accounted for: 0:00:00
                                        Number of                                Number of
                                     applications    Time spent    Percentage        rules
        -------------------------  --------------  ------------  ------------  -----------
        verify atoms                            6       0:00:00            1%            2
        row and column separation               4       0:00:00            2%            -
        obstruction transitivity                4       0:00:00            1%            -
        has specification                       3       0:00:00            1%            -
        factor                                  4       0:00:00            5%            2
        point corroboration                     2       0:00:00            2%            -
        point insertion                         2       0:00:00            3%            1
        add rule                               17       0:00:00            6%            -
        requirement placement                   2       0:00:00           25%            4
        row and column placement                2       0:00:00           47%            8

    ClassDB status:
    	Total number of combinatorial classes found is 7
    	is_empty check applied 6 time. Time spent: 0:00:00
    Queue status (currently on level 2):
        Queue              Size
        ---------------  ------
        working              12
        current (set 1)       2
        next                  0
    	The size of the current queues at each level: 1, 5
    RuleDB status:
                                                   Total number
        ---------------------------------------  --------------
        Combinatorial rules                                   7
        Equivalence rules                                     2
        Combintorial rules up to equivalence                  4
        Strategy verified combinatorial classes               2
        Verified combinatorial classes                        5
        combinatorial classes up to equivalence               4
    	Called find equiv path 3 times, for total time of 0.0 seconds.

    Memory Status:
        ---------------------------------  ---------
        OS Allocated                       345.2 MiB
        Current Memory Used                161.2 MiB
        Current Memory Allocated           225.8 MiB
        Current JIT Memory Used              8.5 MiB
        Current JIT Memory Allocated         9.0 MiB
        Peak Memory Used                   263.6 MiB
        Peak Memory Allocated Memory Used  267.8 MiB
        ---------------------------------  ---------
    	Total Garbage Collection Time: 0:00:11
    Specification found has 4 rules
[I 230122 11:01:31 specification:388] The system of 3 equations
    root_func := F_0(x, k_0):
    eqs := [
    F_0(x, k_0) = F_1(x)*F_2(x, k_0),
    F_1(x) = x*(F_1(x) + 1)**2,
    F_2(x, k_0) = -1/(k_0*x - 1)
    ]:
    count := [0, 0, 0, 0, 0, 0, 0]:
[I 230122 11:01:31 specification:389] Solving...
[I 230122 11:01:33 specification:400] Checking initial conditions for: (2*x - sqrt(1 - 4*x) - 1)/(2*x*(k_0*x - 1))
ic: [0, 0, 0, 0, 0, 0, 0]
[I 230122 11:01:34 specification:400] Checking initial conditions for: (2*x + sqrt(1 - 4*x) - 1)/(2*x*(k_0*x - 1))
ic: [0, 0, 0, 0, 0, 0, 0]
expansion: [0, 1, k_0 + 2, k_0**2 + 2*k_0 + 5, k_0**3 + 2*k_0**2 + 5*k_0 + 14, k_0**4 + 2*k_0**3 + 5*k_0**2 + 14*k_0 + 42, k_0**5 + 2*k_0**4 + 5*k_0**3 + 14*k_0**2 + 42*k_0 + 132]
---------------------------------------------------------------------------
IncorrectGeneratingFunctionError          Traceback (most recent call last)
Cell In[1], line 16
     14     T.classqueue.add(T.classdb.get_label(tiling))
     15 spec = T.auto_search()
---> 16 print(spec.get_maple_equations())

File ~/Dropbox/Research/Active/2017-44-ATRAP/repos/comb_spec_searcher/comb_spec_searcher/specification.py:417, in CombinatorialSpecification.get_maple_equations(self, check)
    413 def get_maple_equations(self, check: int = 6) -> str:
    414     """
    415     Convert the systems of equations to version that can be copy pasted to maple.
    416     """
--> 417     eqs = tuple(self.get_equations())
    418     root_func = self.get_function(self.root)
    419     initial_conditions = self.get_initial_conditions(check)

File ~/Dropbox/Research/Active/2017-44-ATRAP/repos/comb_spec_searcher/comb_spec_searcher/specification.py:333, in CombinatorialSpecification.get_equations(self)
    331 try:
    332     if isinstance(rule, VerificationRule):
--> 333         eq = rule.get_equation(self.get_function, funcs)
    334     else:
    335         eq = rule.get_equation(self.get_function)

File ~/Dropbox/Research/Active/2017-44-ATRAP/repos/comb_spec_searcher/comb_spec_searcher/strategies/rule.py:1152, in VerificationRule.get_equation(self, get_function, funcs)
   1144 def get_equation(
   1145     self,
   1146     get_function: Callable[[CombinatorialClassType], Function],
   1147     funcs: Optional[Dict[CombinatorialClassType, Function]] = None,
   1148 ) -> Eq:
   1149     lhs_func = get_function(self.comb_class)
   1150     return Eq(
   1151         lhs_func,
-> 1152         self.strategy.get_genf(self.comb_class, funcs).subs({var("F"): lhs_func}),
   1153     )

File ~/Dropbox/Research/Active/2017-44-ATRAP/repos/comb_spec_searcher/comb_spec_searcher/strategies/strategy.py:736, in VerificationStrategy.get_genf(self, comb_class, funcs)
    734 if not self.verified(comb_class):
    735     raise StrategyDoesNotApply("The combinatorial class is not verified")
--> 736 return self.get_specification(comb_class).get_genf()

File ~/Dropbox/Research/Active/2017-44-ATRAP/repos/comb_spec_searcher/comb_spec_searcher/specification.py:409, in CombinatorialSpecification.get_genf(self, check)
    407     if expansion == initial_conditions:
    408         return sympy.simplify(genf)
--> 409 raise IncorrectGeneratingFunctionError(
    410     "Failed to compute the generating function for the specification."
    411 )

IncorrectGeneratingFunctionError: Failed to compute the generating function for the specification.

```

File: tilings/strategies/verification.py [hunk 1]
```diff
@@ -280,24 +343,62 @@ def formal_step(self) -> str:
...
+        terms = super().get_terms(comb_class, n)
+        if comb_class.requirements:
+            if len(comb_class.requirements) == 1:
+                comp_spec = self.get_complement_spec(comb_class.remove_assumptions())
+            else:
+                raise NotImplementedError(
+                    "Not implemented counting for one by one with two or more reqs"
+                )
+            comp_terms = comp_spec.get_terms(n)
+            terms = Counter({a: b - comp_terms[a] for a, b in terms.items()})
```
Comment (jaypantone) [inline tilings/strategies/verification.py:355]: Trying to understand this part. Going into this line, won't terms just be a dict with a single entry, key is tuple() and value is a number?

**Review by jaypantone — COMMENTED** [review]
Really great stuff here.  A few thoughts:
 - I think we should find a way to display something like F_123(x) instead of NOTIMPLEMENTED, which will look bad on permpal.
 - We need to discuss the case of a requirement of length 3 or bigger. Rejection sampling could be dangerous and get us stuck in an infinite loop. Also not sure what the complement spec is going to be in these situations.

**Comment by christianbean** [comment]
> Here is a snippet that works on develop but fails on this branch. I'm digging into it further.
> 
> ```python
> import requests
> from comb_spec_searcher import CombinatorialSpecification
> from comb_spec_searcher.rule_db import RuleDBForest
> from tilings.tilescope import TileScope, TileScopePack
> 
> uri = "https://permpal.com/perms/raw_data_json/basis/0123_0231"
> 
> data = requests.get(uri).json()
> spec = CombinatorialSpecification.from_dict(data["specs_and_eqs"][0]["spec_json"])
> pack = TileScopePack.from_dict(data["specs_and_eqs"][0]["pack_json"])
> 
> T = TileScope(spec.root, pack, ruledb=RuleDBForest())
> for tiling in spec.comb_classes():
>     T.classqueue.add(T.classdb.get_label(tiling))
> spec = T.auto_search()
> print(spec.get_maple_equations())
> ```

This has been fixed by the last commit.

File: tilings/strategies/verification.py (hunk 1 above)
Comment (christianbean) [reply tilings/strategies/verification.py:358]: With the latest two commits this is now true. I'll simplify this line.

**Comment by christianbean** [comment]
> Really great stuff here. A few thoughts:
> 
> * I think we should find a way to display something like F_123(x) instead of NOTIMPLEMENTED, which will look bad on permpal.

So the reason this is hard is that we have to make it compatible with `comb_spec_searcher.utils.sympy_expr_to_maple`. If we add a hard fix there to look for an expression, e.g, F_Av(0123), then it's possible. It does however need this change on the CSS function. 

The sympy equation for the rule and so `CombinatorialSpecification.get_pretty_equations`, will work fine out the box. Only `CombinatorialSpecification.get_maple_equations` will throw an error.. 

> * We need to discuss the case of a requirement of length 3 or bigger. Rejection sampling could be dangerous and get us stuck in an infinite loop. Also not sure what the complement spec is going to be in these situations.

It is possible but unlikely to get infinite loops, and so I think this is a nice feature to be had.

If your class is Av(B) intersect Co(C) then the complement spec is Av(B union C). 

Note, the rejection sampling doesn't use the complement spec, just checks tiling containment.

**Pull request merged**